File: src/main/kotlin/com/mobaassistant/service/TeamCoordinationService.kt
package com.mobaassistant.service

import com.mobaassistant.model.GameState
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import org.springframework.stereotype.Service

@Service
class TeamCoordinationService {
    private val quickCommands = mapOf(
        "gather" to "Group up for objective",
        "retreat" to "Fall back - enemy missing",
        "push" to "Push towers while ahead",
        "defend" to "Defend base - enemy pushing"
    )

    fun generateCallouts(gameState: GameState): List<CoordinationCommand> {
        val commands = mutableListOf<CoordinationCommand>()

        if (isTeamfightImminent(gameState)) {
            commands.add(CoordinationCommand(
                type = CommandType.TEAMFIGHT,
                message = "Prepare for team fight",
                priority = Priority.HIGH
            ))
        }

        gameState.currentObjective?.let { objective ->
            commands.add(CoordinationCommand(
                type = CommandType.OBJECTIVE,
                message = "Setup for $objective",
                priority = Priority.MEDIUM
            ))
        }

        return commands
    }

    fun getQuickCommand(situation: String): String {
        return quickCommands[situation.lowercase()] ?: "Invalid command"
    }

    fun getRoleSpecificCallouts(role: String): List<String> {
        return when (role.lowercase()) {
            "tank" -> listOf(
                "Initiate when team is ready",
                "Protect carries in teamfight"
            )
            "marksman" -> listOf(
                "Focus highest priority target",
                "Stay behind frontline"
            )
            "assassin" -> listOf(
                "Look for pick-off opportunities",
                "Split push side lanes"
            )
            else -> emptyList()
        }
    }

    fun observeTeamCoordination(gameState: GameState): Flow<List<CoordinationCommand>> = flow {
        while (true) {
            emit(generateCallouts(gameState))
            kotlinx.coroutines.delay(1000)
        }
    }

    private fun isTeamfightImminent(gameState: GameState): Boolean {
        // Implement teamfight prediction logic
        return gameState.currentObjective != null
    }
}

enum class CommandType {
    TEAMFIGHT, OBJECTIVE, ROTATION, GENERAL
}

enum class Priority {
    LOW, MEDIUM, HIGH
}

data class CoordinationCommand(
    val type: CommandType,
    val message: String,
    val priority: Priority,
    val timestamp: Long = System.currentTimeMillis()
)

File: app/services/team-coordination.service.ts
import { Observable } from '@nativescript/core';
import { GameState } from '../models/game-state.model';
import { CoordinationCommand } from '../models/coordination.model';

export class TeamCoordinationService extends Observable {
  private gameState: GameState | null = null;
  private readonly QUICK_COMMANDS = new Map<string, string>([
    ['gather', 'Group up for objective'],
    ['retreat', 'Fall back - enemy missing'],
    ['push', 'Push towers while ahead'],
    ['defend', 'Defend base - enemy pushing']
  ]);

  generateCallouts(): CoordinationCommand[] {
    if (!this.gameState) return [];

    const commands: CoordinationCommand[] = [];
    const { allyTeam, currentObjective } = this.gameState;

    // Generate situation-specific callouts
    if (this.isTeamfightImminent()) {
      commands.push({
        type: 'teamfight',
        message: 'Prepare for team fight',
        priority: 'high'
      });
    }

    if (currentObjective) {
      commands.push({
        type: 'objective',
        message: `Setup for ${currentObjective}`,
        priority: 'medium'
      });
    }

    return commands;
  }

  getQuickCommand(situation: string): string {
    return this.QUICK_COMMANDS.get(situation) || 'Invalid command';
  }

  getRoleSpecificCallouts(role: string): string[] {
    const callouts: string[] = [];

    switch (role.toLowerCase()) {
      case 'tank':
        callouts.push('Initiate when team is ready');
        callouts.push('Protect carries in teamfight');
        break;
      case 'marksman':
        callouts.push('Focus highest priority target');
        callouts.push('Stay behind frontline');
        break;
      case 'assassin':
        callouts.push('Look for pick-off opportunities');
        callouts.push('Split push side lanes');
        break;
    }

    return callouts;
  }

  private isTeamfightImminent(): boolean {
    if (!this.gameState) return false;
    
    // Check for teamfight indicators:
    // - Multiple enemies grouped
    // - Objective spawning
    // - Key abilities available
    return true; // Simplified
  }

  updateGameState(newState: GameState): void {
    this.gameState = newState;
    this.notifyPropertyChange('gameState', newState);
  }
}